text_encoder.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/uber-go/atomic"
+)
+
+var textPool = sync.Pool{New: func() interface{} {
+	return &textEncoder{
+		bytes:  make([]byte, 0, _initialBufSize),
+		nested: atomic.NewBool(false),
+	}
+}}
+
+type textEncoder struct {
+	bytes   []byte
+	timeFmt string
+	nested  *atomic.Bool
```

Comment:
The encoder doesn't need to be safe for concurrent use when adding fields.

The only slightly confusing piece is that `Clone` and `WriteEntry` may be called concurrently, as described here:
https://godoc.org/github.com/uber-go/zap#Encoder

Since this will only be modified when adding fields, no need for atomic here.